// @serwist/build/dist/chunks/inject-manifest-DNqDY-04.js
import { t as __exportAll } from "./rolldown-runtime-CzwdidQP.js";
import { i as basePartial, r as requiredGlobDirectoryPartial, t as globPartial } from "./glob-B4_meZ_o.js";
import { z } from "zod";
//#region src/schema/sw-dest.ts
const optionalSwDestPartial = z.strictObject({ swDest: z.string().optional() });
const requiredSwDestPartial = z.strictObject({ swDest: z.string() });
//#endregion
//#region src/schema/inject-manifest.ts
var inject_manifest_exports = /* @__PURE__ */ __exportAll({
	baseInjectPartial: () => baseInjectPartial,
	injectManifestOptions: () => injectManifestOptions
});
const baseInjectPartial = z.strictObject({
	injectionPoint: z.string().default("self.__SW_MANIFEST"),
	swSrc: z.string()
});
const injectManifestOptions = z.strictObject({
	...basePartial.shape,
	...globPartial.shape,
	...baseInjectPartial.shape,
	...requiredSwDestPartial.shape,
	...requiredGlobDirectoryPartial.shape
});
//#endregion
export { requiredSwDestPartial as a, optionalSwDestPartial as i, injectManifestOptions as n, inject_manifest_exports as r, baseInjectPartial as t };
